### PR TITLE
Big-ish emote handling refactor

### PR DESCRIPTION
**Pull request by RAnders00**
Don't merge yet, bingo is still missing + i need to test some more changes kkOna

TODO

- [x] Potentially make `Emote` and `EmoteInstance` classes
- [x] Test emotes on screen module
- [x] Test showemote module
- [x] Test CLR behaviour of:
    - [x] 1x-only emote (-> TriEasy)
    - [x] 2x-only emote (-> Hahaa)
    - [x] 3x-only emotes
- [x] Test emote quest module
- [x] Add unit test for `fill_in_url_scheme`
- [x] Add unit test for `iterate_split_with_index`
- [x] Test that regular refresh (every 1 hour) works correctly
- [x] Reimplement bingo
- [x] Should `etmrecord`/`epmrecord` be re-added? There was only a getter, not a setter for it

### CODE REVIEW

File: alembic/versions/5f746af0a82d_redis_delete_emote_hashes.py [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+def upgrade():
+    with RedisManager.pipeline_context() as redis:
+        redis.delete('{streamer}:emotes:ffz_channel_emotes'.format(streamer=streamer))
+        redis.delete('{streamer}:emotes:bttv_channel_emotes'.format(streamer=streamer))
+        redis.delete('global:emotes:ffz_global')
+        redis.delete('global:emotes:bttv_global')
+
+def downgrade():
+    with RedisManager.pipeline_context() as redis:
+        redis.delete('{streamer}:emotes:twitch_channel_emotes'.format(streamer=streamer))
```
Comment (RAnders00) [inline alembic/versions/5f746af0a82d_redis_delete_emote_hashes.py]: todo remove `_emotes` suffix here

**Comment by RAnders00** [comment]
Caveats so far: `$(tb:etmrecord)` is gone for now, can be readded if needed

File: pajbot/managers/emote.py [hunk 2]
```diff
@@ -1,422 +1,439 @@
...
+        for emote_code, obj in emote_counts.items():
+            self.epm_incr(emote_code, obj["count"])
 
     def epm_incr(self, code, count):
-        if code in self.epm:
-            self.epm[code] += count
-        else:
-            self.epm[code] = count
+        self.epm[code] = self.epm.get(code, 0) + count
+        # TODO if we want to add epm records back, do it here
```
Comment (RAnders00) [inline pajbot/managers/emote.py]: xd

File: pajbot/managers/handler.py [hunk 3]
```diff
@@ -64,9 +64,6 @@ def init_handlers():
         # on_user_resub(user, num_months)
         HandlerManager.create_handler("on_user_resub")
 
-        # send_whisper(user, message)
-        HandlerManager.create_handler("send_whisper")
```
Comment (RAnders00) [inline pajbot/managers/handler.py]: Was replaced by invocations of bot.whisper

File: pajbot/managers/handler.py [hunk 4]
```diff
@@ -52,9 +55,6 @@ def init_handlers():
         # on_duel_complete(winner, loser, points_won, points_bet)
         HandlerManager.create_handler("on_duel_complete")
 
-        # on_user_gain_tokens(user, tokens_gained)
-        HandlerManager.create_handler("on_user_gain_tokens")
```
Comment (RAnders00) [inline pajbot/managers/handler.py]: Not used at all (Was never emitted)

File: static/scripts/clr.overlay.js [hunk 5]
```diff
@@ -39,49 +39,43 @@ function add_random_box({color}) {
...
+        newdiv.appendTo('body');
+        newdiv.animate({
+            opacity: opacity / 100
+        }, 500);
+        setTimeout(() => {
             newdiv.animate({
-                opacity: opacity / 100
-            }, 500);
+                opacity: 0,
+            }, 1000);
```
Comment (RAnders00) [inline static/scripts/clr.overlay.js]: idk what happened here

**Comment by pajlada** [comment]
etmrecord was never implemented, so removing it is fair

**Comment by RAnders00** [comment]
It's possible to do etmrecord, I saw some redis eval functions (lua code you run in redis) that's basically `ZADD` but only if the new value is greater.

**Comment by pajlada** [comment]
probably worth exploring, possibly as a separate PR?
if you feel like it's something you might get to, you could probably just leave the etmrecord replacements in

**Comment by RAnders00** [comment]
Good commands to run after this update:

```none
!add command epmrecord @$(source:username_raw), $(1) per minute record is $(epmrecord;1).
!edit command epm @$(source:username_raw), $(1) is currently being used $(epm;1) times per minute.
```

**Pull request merged**